cclunit-framework-source/doc/CCLUTMOCKING.md

```diff
@@ -0,0 +1,268 @@
+  
+@param originalName  
+&nbsp;&nbsp;&nbsp;&nbsp;The object that is mocked.  
+  
+Example:  
+call cclutRemoveMockImplementation("uar_get_code_by")
+
+**cclutRemoveAllMockImplementations(null)**
+
+Removes all mock implementations.  
+  
+Example:  
+call cclutRemoveAllMockImplementations(null)
+
+**cclutExecuteProgramWithMocks(programName = vc, params = vc, namespace = vc)**
+
+Executes a program with all mocks currently added through cclutAddMockImplementation() and cclutCreateMockTable(). programName is required.  params is a string parameter to be sent directly to the program, so all commas and string delimiters must be specified.  If namespace is omitted, it will default to the PUBLIC namespace.  
+  
+Example:  
+call cclutExecuteProgramWithMocks("ccl_my_program", "\^MINE^, 1.0, ^string parameter^", "MYNAMESPACE")
+
+**cclutRemoveAllMocks**
+
+Removes all mock implementations and mock tables that have been added through the cclutAddMockImplementation() and cclutCreateMockTable() APIs.  This should be called at the completion of a test suite to clean up all mocks.  
+  
+Example:  
+call cclutRemoveAllMocks(null)
+
+## Implementation Notes
+1. cclutRemoveAllMocks should be called as part of the teardown for all tests.  The framework will attempt to clean up any outstanding mocks, but it is good practice to explicitly remove any mocks to ensure that no mocked tables remain in the Oracle instance.
+
+2. The mocked items created through cclutCreateMockTable() and cclutAddMockImplementation will not be applied to children script called from the script-under-test.  Some alternatives would be to mock the child script to return the appropriate data or to mock the child script to execute the real script applying the mocked tables and implementations.
+
+3. The mocked items created through cclutCreateMockTable() and cclutAddMockImplementation will not be applied to statements executed through "call parser()" commands.  An alternative would be to mock the parser() call to validate the correct information is supplied, then perform the appropriate mock versions of the actions the statement would normally perform.
+
+4. The mocking API calls cannot be used from reportwriter sections.  Alternatives would be dependent on the use-case, but take, for example, a dummyt used with a record structure and a call to cclutAddMockData within the detail section in order to add data based on a record structure.  Instead, a FOR loop construct could be leveraged outside the context of a reportwriter section to iterate over the record structure.
+
+5. Mocking record structures with a call to tdbexecute is unsupported under certain conditions, specifically if a call to free the record structure is made just prior to calling tdbexecute.  If the scenario is truly necessary for a test, the best alternative is to separate the freeing of the record structure and the call to tdbexecute in different subroutines and test the subroutines independently of each other.
```

Comment:
> Mocking the tdbexecute "reply_to" entity is unsupported.... a call to free the "reply_to" entity is made...
... define and use a subroutine for freeing the "reply_to" entity within the script and use a mock for that subroutine which does not actually perform the free.

I'll make the changes for the tdbexecute wording.

> The following thing might deserve mention in this section since there are already two other cases that should be handled in the exactly same way. Hard-coded commit statements are obnoxious in general but especially so for testing. The only way to test that a hard-coded commit does or does not get performed at appropriate/inappropriate times or to avoid a call to one is to create and use a "doCommit" subroutine for performing the commit within the sut and then mock doCommit to spy on and/or avoid the commit call. Observe that using with replace("COMMIT", whatever) does not work.

Do you have a more specific example to help me understand the concern?

I'm trying to think of use cases for testing commits, and only two thoughts are coming to mind - one where a subroutine does some DML and fires a commit and one where a subroutine does some DML and does not fire a commit (and even possibly issues a rollback though that doesn't particularly matter).

If I assume the following subroutine to test:
```
subroutine test_subroutine()
    update into person p
    set
    	   p.name_last = "new last name",
    	   p.name_first = "new first name",       	
           p.birth_dt_tm = cnvtdatetime(curdate, curtime3)
    where p.person_id = 1

    commit
end
```  

This could be tested by having the unit test issue a "rollback" command immediately after calling the subroutine under test.  If the changes expected are still present in the (mocked) table, then the commit has been executed.  Removing the commit line would cause the test to fail.

Similarly, if the commit _shouldn't_ be there (and the unit test is testing that), it can check the table after executing the subroutine under test.  The changes should be present.  Then it can issue a rollback and check if the changes have disappeared.  If they have not, they must have been committed and the test can fail.

Reply:
Before getting too deep in the weeds on this, note that I made this comment only because this section is starting to feel like unit testing guidance. Perhaps there should just be pages dedicated to guidance, tips, tricks, etc. where some of this would go. 

I agree that a clever use of mock tables and logic will handle happy path of the given example, but what about the exception path.  If an error occurs during the update, there will be no data regardless of whether a commit or rollback was or was not performed.  Perhaps there is an error check between the update and the commit with the expectation that the error check will perform a rollback and route control past the commit if an error occurs. Perhaps there are even several independent updates involved.  

The goal of a test is to prove something did or did not happen not just to make a crafty guess.  Suppose you want to test that function A calls \[does not call\] function/script B in a given scenario. Wouldn't you just mock B to say "I've been called" and test for that? The point here is that it is impossible to mock commit or rollback. The alternative is to wrap them in something that can be mocked.   ....not to mention there is ample reason to avoid calling any commit in a unit test.  Who knows what might be getting committed.

Reply:
> Before getting too deep in the weeds on this, note that I made this comment only because this section is starting to feel like unit testing guidance. Perhaps there should just be pages dedicated to guidance, tips, tricks, etc. where some of this would go.

I think that might be fine to have a separate section.  My purpose in having this (and titling it "Implementation Notes") was to provide information to consumers without them needing to look at the implementation to determine whether something would be viable or not.  As an example, the cclutAddMockData subroutine can't be leveraged from reportwriter sections because of its usage of DML.  Someone looking at just the subroutine name might not realize that (since you can call subroutines that _don't_ use DML from reportwriter sections just fine), but I didn't want to have "implementation details" per se within the API doc, so I felt like this separate section was the best way to represent them.  If we want to have a separate section (or incorporate that other guidance into here), I'm fine with that, but I'd like to get a listing of those additional things we might add.  I've incorporated (locally) your previous suggestion that the objects/tables get normalized to all-uppercase, but I'm still not sold on the commit stuff.

> I agree that a clever use of mock tables and logic will handle happy path of the given example, but what about the exception path. If an error occurs during the update, there will be no data regardless of whether a commit or rollback was or was not performed. Perhaps there is an error check between the update and the commit with the expectation that the error check will perform a rollback and route control past the commit if an error occurs. Perhaps there are even several independent updates involved.

I'm still having trouble visualizing why this scenario necessitates a separate subroutine for the commit action and how specifically having a separate subroutine for commit actions somehow mitigates the concern.  Specifically, this line "If an error occurs during the update, there will be no data regardless of whether a commit or rollback was or was not performed."  This would still be true even if the hypothetical commit subroutine was mocked to do nothing since the error occurred on the update.

> The goal of a test is to prove something did or did not happen not just to make a crafty guess. Suppose you want to test that function A calls [does not call] function/script B in a given scenario. Wouldn't you just mock B to say "I've been called" and test for that? The point here is that it is impossible to mock commit or rollback. The alternative is to wrap them in something that can be mocked. ....not to mention there is ample reason to avoid calling any commit in a unit test. Who knows what might be getting committed.

It sounds like we might have disagreements about the nature of unit testing, which is fine, I'll structure it however you prefer, but I just want to confirm that's the case as I don't want to structure my arguments based on differing premises.  In your example where B is already a function, then yes, I'd just mock it, but if it's not already a function (i.e. B's contents are in A), then I don't believe it's the responsibility of the framework to dictate to developers how their tests should be structured.  The framework should be flexible enough to support a wide variety of testing, including not having every instance of commit/rollback segregated to its own subroutine.

In cases where it's not flexible enough (such as the "free record"/tdbexecute issue), I think it's worthwhile to bring up in this section, or to bring up instances where the API is not fully honoring its stated purpose (such as parser() or child scripts), but it doesn't seem like the commit guidance meets that criteria.  I was challenging your initial premise that the "only way" to test the commit was to have it be in a separate subroutine.  I don't believe that to be true (at least, I'm having difficulty coming up with a scenario hence my request if you had a concrete example), but I'm happy to be proven wrong because I want to make sure that the documentation is robust enough to detail all of those kinds of scenarios.  

To me, the mocking API doesn't change anything about commits/rollbacks (just which tables it's applied to), so even if there were such a situation, it would apply to the current iteration of the framework instead of just now when we add mocking (hence why a separate section might be good).  But if this is the only thing that would go in there, and (at least in my opinion) isn't needed anyway, it doesn't seem appropriate to have a separate section just for it.

All of this is just to present my views, though, so I'll gladly make any changes you request if it's not persuasive enough.

Reply:
> When originally reviewing this section I noticed that there is a whole class of unit testing stumbling blocks which have a common solution - use a helper function to do something in CCL that could otherwise be done directly. The testing framework does not dictate this need (nor should it). The necessity arises because of CCL limitations coupled with a desire to test thoroughly. I am not a fan of changing code to fit a testing framework, but some code is not testable and must be written differently to become testable (regardless the testing framework).

With regards to "there is a whole class of unit testing stumbling blocks which have a common solution - use a helper function to do something in CCL that could otherwise be done directly.", this is not true for the "free record"/tdbexecute.  There is no direct alternative (unless you're aware of one), and the helper function _must_ be used if the consumer wishes to test it.

On a similar note, I added the parser() alternative around using a separate subroutine based on your suggestion, but that's not the reason why the parser() guidance is in this section.  I put the parser() guidance here so that consumers understood that this particular implementation of the mock table API would not replace those instances inside parser().  If one just reads the raw API (with no knowledge of the implementation), they might presume that the mock table API _would_ work with parser() statements.

For those reasons, I view these two pieces of guidance (which are closely related the mocking API/implementation) very differently from the commit guidance you're proposing which has no bearing on the mocking API.

> Here is some additional stuff that could go in the separate section: testable CCL and CCL mocking

In terms of an additional section, these would be good to include as well.  Let me draft something that includes these and the commit guidance to get your thoughts if it would be satisfactory.

> The following test code very clearly demonstrates that commit gets called and rollback does not get called on the happy path and that rollback gets called and commit does not get called on an exception path involving an insert error (ignoring the fact that commit and rollback cannot be mocked in this way). A test that tries to do this in any other way would be less than convincing and/or require unnecessary effort for the reader to become convinced.

I would still contest that the test you have is not testing that commit and rollback are called.  As you mention, the commands can't be mocked in that way, so you'd need to have a commitSubroutine and rollbackSubroutine.  These tests are testing that those _subroutines_ are called, not that the commands are called.  For full coverage, the consumer should also want to test those utility subroutines as well, putting us back where we started.  

I also don't think it would be less convincing if the test itself were to validate the items that are in the table.  In such a way, the test is testing the behavior of the subroutine rather than the specific syntax used in the implementation.

> I personally would be uneasy creating a test for a program that performs a hard-coded commit without mocking the commit call. My issue is that future maintenance could easily alter what actually happens when that test is executed leading to the unexpected commit of db updates performed during test setup or program execution.

I can understand this concern (especially in scenarios where mocked tables _aren't_ utilized), and given your other suggestions for a "tips and tricks" section, I'd be happy to include the commit guidance in such a section.  As mentioned, I'll draft something for review in a later commit.  Do you feel that would be appropriate, or do you still want the commit guidance in the "Implementation Notes" section?

Reply:
I agree with what you are saying. I am just advocating **brief** mention of potential issues here with a link to a tips and tricks section that goes into details.  I don't think very many people will read this section. It is more likely that they will glance at the example and then jump right in.  People who run into issues on their very first tries at mocking stuff will dig a little deeper into the documentation. The others will forget that it exists by the time they start running into these kinds of problems and will be more likely to open an FAQ type document than to reread the documentation looking for answers.

Reply:
Fair enough.  Let's revisit after I've made these changes, and we can discuss further.